kratos/utilities/function_parser_utility.cpp

```diff
@@ -24,25 +27,64 @@
 namespace Kratos
 {
 
-// A check only to check that the x is not part of exp
-inline bool CheckThereIsNotx(const std::string& rString)
+/**
+ * @brief Ternary version of adjacent_find, with padding.
+ * @details Scans a sequence with a 3-window, passing it to a predicate.
+ * The window is padded such that every element of the sequence
+ * is the middle element of the window once.
+ * @return Returns the iterator to the middle of the first window to
+ * fulfill the predicate.
+ */
+template<typename ForwardIterator, typename TernaryPredicate>
+ForwardIterator PaddedWindowFind(
+    ForwardIterator Begin,
+    ForwardIterator End,
+    typename std::iterator_traits<ForwardIterator>::value_type Padding,
+    TernaryPredicate&& Pred
+    )
 {
-    const auto first_x = rString.find(std::string("x"));
-    if (first_x != std::string::npos) {
-        const std::string exp_string = "exp";
-        const char e_char = exp_string[0];
-        const char x_char = exp_string[1];
-        const char p_char = exp_string[2];
-        for(std::size_t i = first_x; i < rString.size(); ++i) {
-            if(rString[i] == x_char) {
-                if (!(rString[i - 1] == e_char && rString[i + 1] == p_char)) {
-                    return false;
-                }
-            }
-        }
+    // Edge case: length 0
+    if(Begin == End) {
+        return End;
+    }
+    
+    auto wmiddle = std::next(Begin); // Middle of the window
```

Comment:
> The window is padded such that every element of the sequence is the middle element of the window once.

Minor thing: this is a bit confusing because ```wmiddle``` will never point to ```Begin```, but I see the necessary check is performed in line 58 so I think the behaviour is correct.

Reply:
Yeah, the behaviour is correct (as you pointed out about line 58). I could make it clearer but that would require extra copies and extra `std::advance`s.

Reply:
Yeah I'd say don't worry about it, it just takes a minute to figure out what's going on.